rfcs/text/refactoring-date-component.md

```diff
@@ -0,0 +1,399 @@
+
+```js
+  const renderWeekdayElement = (weekdayElementProps) => {
+    const { className, weekday } = weekdayElementProps;
+    const weekdayLong = localize.day(weekday);     
+    const weekdayShort = localize.day(weekday, { width: "abbreviated" });
+    
+    return (
+      <Weekday className={className} title={weekdayLong}>
+        <abbr title={weekdayLong}>{weekdayShort}</abbr>
+      </Weekday>
+    );
+  }}
+```
+
+```js
+  const months = Array.from({ length: 12 }).map((_, i) => localize.month(i))
+```
+
+We will no longer maintain the last valid value in the `DatePicker`, if a user inputs a value that cannot be parsed to a valid date, the `selectedDate` prop will be set to `undefined`.  We will also keep the same implementation the current component has with regards to rendering the `DatePicker` in a `portal` using the existing `Popover` component. To support additional features we will expose the `react-day-picker` interface via the `pickerProps`. For example, the following [feature request](https://github.com/Sage/carbon/issues/4275) can be supported this by utilising the [captionElement](https://react-day-picker.js.org/examples/elements-year-navigation) render prop, although we would implement this via a boolean prop and handle the the composition internally ustilising Carbon's existing `Select` component: below is an example of how this can be achieved.
+
+```js
+const YearMonthForm = ({
+  date = new Date(),
+  months,
+  onChange,
+  minDate,
+  maxDate,
+}) => {
+  const currentYear = date.getFullYear();
+  const toMonth = maxDate || new Date(currentYear + 10, 11);
+  const fromMonth = minDate || new Date(currentYear, 0);
+  const years = [];
+  
+  for (let i = fromMonth.getFullYear(); i <= toMonth.getFullYear(); i += 1) {
+    years.push(i);
+  }
+
+  const handleChange = (ev) => {
+    const { year, month } = ev.target.form;
+
+    const yearValue = year?.value || date.getFullYear();
+    const monthValue = month?.value || date.getMonth();
+    const dayValue = date.getDate();
+
+    onChange(new Date(yearValue, monthValue, dayValue));
+  };
+
+  return (
+    <form>
+      <Select name="month" onChange={handleChange} value={date.getMonth()}>
+        {months.map((month, i) => (
+          <Option key={month} value={i}>
+            {month}
+          </Option>
+        ))}
+      </select>
+      <Select name="year" onChange={handleChange} value={date.getFullYear()}>
+        {years.map((year) => (
+          <Option key={year} value={year}>
+            {year}
+          </Option>
+        ))}
+      </select>
+    </form>
+  );
+}
+```
+
+```jsx
+  // date/__internal__/date-picker/date-picker.component.js
+  import {
+    localeMap,
+    isValidDate,
+    parseToDate
+  } from "./__internal__/date-utils";
+
+  const DatePicker = ({
+    inputValue,
+    selectedDate,
+    setSelectedDate,
+    onDayClick,
```

Comment:
So I understand that you can either:
1. Select Date using Picker and your mouse
2. Type in the date and ignore the picker

Shouldn't DatePicker have `onChange` as `onDayClick` is not sufficient?

Reply:
Yes so the typical pattern would be what you've described, we also intend to make the picker keyboard accessible as well so a user will be able to select a date using tabbing, arrow keys and then enter to select the date. There's no need to pass an onChange to the picker for this to work but the `containerProps` interface supports passing the event callbacks if we want. I've updated the POC to include an example of the keyboard support working

https://codesandbox.io/s/poc-of-rfc-proposal-ctdfw?file=/src/index.js:3195-3197

Reply:
sweet, sounds great. Great work guys, I know first hand what this Draft means as I was "suffering" while implementing an old Carbon DateInput component. This proposed changes are very well thought through!

Reply:
Thanks for the feedback and taking the time to read it @damianOfKyan 😄